### ISSUE TITLE
Cannot Reproduce Offline Training Results in Table 1

### ISSUE THREAD

**Issue by m90113808** [YifanYang1995/GOODRL#1]
Dear authors,  

Thank you for open-sourcing GOODRL and providing the code & data. We tried to reproduce the offline results in **Table 1** using the exact instructions in the [README](https://github.com/YifanYang1995/GOODRL/blob/master/README.md), but could not match the reported performance.  

In the **<6 × 4, 5.4, \*> scenarios**, the paper reports mean flowtime ≈ **277**, but in our runs:  

- Warm-up stage: stable around 298 (as expected)  
- Training stage: mean flowtime quickly **diverges** and does not converge  

Excerpt from our training log (full log in [step2.log](https://github.com/user-attachments/files/22450172/step2.log)):  

```
Vlidation at update-0: mean_flowtime_deterministic: 298.383075+/-46.569598 mean_Entropy: 2.107124+/-0.036911 record: 298.383075 time_elapsed: 0.031238
Episode-1: ep_meanFlowTime: 495.401568 all_loss: 1.286482 p_loss: -1.481019 e_loss: 1.982751 v_loss: 2.572964 v_mre: 90.868090 grad_changes: 0.000000 time_elapsed: 0.0573
...
Episode-200: ep_meanFlowTime: 400.252887 all_loss: 0.835384 p_loss: -1.163283 e_loss: 2.035066 v_loss: 1.670768 v_mre: 88.109463 grad_changes: 0.000000 time_elapsed: 4.9378
Vlidation at update-200: mean_flowtime_deterministic: 298.384307+/-46.570831 mean_Entropy: 2.107048+/-0.036862 record: 298.382860 time_elapsed: 4.9565
Episode-201: ep_meanFlowTime: 365.211668 all_loss: 1.842912 p_loss: -1.127642 e_loss: 2.295801 v_loss: 1.430541 v_mre: 87.363136 grad_changes: 0.073420 time_elapsed: 4.9817
...
Episode-1200: ep_meanFlowTime: 2051.066130 all_loss: 55.116712 p_loss: -6.757592 e_loss: 2.816375 v_loss: 96.718240 v_mre: 96.630242 grad_changes: 0.000000 time_elapsed: 27.2327
Vlidation at update-1200: mean_flowtime_deterministic: 2357.957823+/-843.660011 mean_Entropy: 2.727679+/-0.083880 record: 298.382860 time_elapsed: 27.2564
```

Could you please advise:  
1. Are there any **critical hyperparameters or steps** not listed in the README?  
2. Could you share a sample config/log to help reproduce results?  

Many thanks for your guidance and for the valuable contribution!

**Comment by YifanYang1995** [comment]
Dear User,

Thank you for your interest in our work and for running the GOODRL code. We appreciate you sharing the detailed logs of your reproduction attempt.

Please note that the results reported in **Table 1** of our paper are the **test set performance** of the best policy, while the logs you shared show **validation set performance** during training. This explains why there's a discrepancy between your validation metrics and our reported test results.


For your **first question**, here are the exact parameters we used for training on the `<6 × 4, 5.4, 30>` scenario:

```
python step2.py --algo_seed 3 --env_seed 0 --wf_size S --wf_num 30 --num_envs 4 --valid_num 30\
    --vm_types 6 --each_vm_type_num 4 --arr_rate 5.4 --log_interval 20 --entloss_coef 0 --vloss_coef 0.5\
    --require_estimated_features 1 --require_mean 0 --require_undirected 1 --remove_completed 0\
    --hidden_dim 128 --batch_size 64 --gnn_layers 2 --atten_layers 1 --mlp_layers 4 --heads 1 --max_updates 1000\
    --lr_a 0.0003 --lr_c 0.001 --n_epochs 1 --require_clip_value 4 --normalize_rewards 1000 --warmup_critic 200
```

For your **second question**, we're attaching two of our training logs for reference:
- [**`<6 × 4, 5.4, 30>`** scenario](https://github.com/user-attachments/files/22740468/slurm-49832298_3.log)
```
Vlidation at update-0: mean_flowtime_deterministic: 318.1444+/-48.7547	 mean_Entropy: 2.1258+/-0.0409	 record: 318.1444	 time_elapsed: 0.1466
Episode-1 with gradient: 0.0000+/-0.0000 
Episode-1: ep_meanFlowTime: 441.2057	 p_loss: -1.4345	 e_loss: 1.9748	 v_loss: 0.6426	 all_loss: 0.3213	 time_elapsed: 0.19
...
Episode-200 with gradient: 0.0000+/-0.0000 
Episode-200: ep_meanFlowTime: 435.8426	 p_loss: -0.0020	 e_loss: 1.9685	 v_loss: 0.0743	 all_loss: 0.0372	 time_elapsed: 12.88
Vlidation at update-200: mean_flowtime_deterministic: 318.1444+/-48.7547	 mean_Entropy: 2.1256+/-0.0405	 record: 318.1444	 time_elapsed: 12.9823
...
Episode-1200 with gradient: 0.0037+/-0.0125 
Episode-1200: ep_meanFlowTime: 297.5223	 p_loss: -0.0011	 e_loss: 0.6207	 v_loss: 0.0041	 all_loss: 0.0031	 time_elapsed: 77.89
Vlidation at update-1200: mean_flowtime_deterministic: 307.6949+/-43.4414	 mean_Entropy: 0.6359+/-0.0196	 record: 305.3595	 time_elapsed: 78.0027
```
- [**`<5 × 5, 5.4, 30>`** scenario](https://github.com/user-attachments/files/22740467/slurm-49965018_4.log)
```
Vlidation at update-0: mean_flowtime_deterministic: 458.4526+/-70.5696	 mean_Entropy: 1.3481+/-0.0342	 record: 458.4526	 time_elapsed: 0.1824
Episode-1 with gradient: 0.0000+/-0.0000 
Episode-1: ep_meanFlowTime: 439.8901	 p_loss: -1.3134	 e_loss: 1.2588	 v_loss: 0.6811	 all_loss: 0.3406	 time_elapsed: 0.24
...
Episode-200 with gradient: 0.0000+/-0.0000 
Episode-200: ep_meanFlowTime: 451.5670	 p_loss: -0.0592	 e_loss: 1.2616	 v_loss: 0.0691	 all_loss: 0.0346	 time_elapsed: 13.49
Vlidation at update-200: mean_flowtime_deterministic: 458.4525+/-70.5697	 mean_Entropy: 1.3481+/-0.0343	 record: 458.4525	 time_elapsed: 13.6253
...
Episode-1200 with gradient: 0.0033+/-0.0202 
Episode-1200: ep_meanFlowTime: 423.8427	 p_loss: -0.0112	 e_loss: 0.2860	 v_loss: 0.0034	 all_loss: 0.0129	 time_elapsed: 70.26
Vlidation at update-1200: mean_flowtime_deterministic: 446.9479+/-65.6370	 mean_Entropy: 0.3274+/-0.0368	 record: 443.9528	 time_elapsed: 70.3499
```
*Note: These logs were generated with our original development version. The GitHub release has been cleaned up for better presentation, which might cause minor differences.*

Looking at your [`step2.log`](https://github.com/user-attachments/files/22450172/step2.log), we identified a critical issue:

> The **`v_mre`** (value mean relative error) metric shows **no decreasing trend** throughout training, indicating that the **value function is not learning properly**. When the value estimates are inaccurate, it severely destabilizes the actor's policy updates, leading to the catastrophic performance divergence you observed.

For the next step, we recommend:
1. **Verify value function training**: Ensure the critic network is properly initialized and receiving gradients
2. **Monitor gradient flow**: Confirm that both actor and critic networks are receiving and applying updates

If you continue to experience issues, we're happy to continue this discussion and help you achieve successful reproduction!

Best regards,
The GOODRL Team

**Comment by m90113808** [comment]
Hi @YifanYang1995,  

Thank you very much for your detailed reply and for sharing the clarifications. We reran the **<6 × 4, 5.4, 30>** experiment using the source code and the exact commands you provided. Unfortunately, our results remain similar to the previous reproduction attempts — we still cannot achieve the stable low mean flowtime on the validation set after around 200 episodes, as shown in your provided log.  

While we understand that the mean flowtime on the **validation set** is not strictly equivalent to that on the **test set**, a model yielding over **2k+ mean flowtime** on the validation set seems unlikely to reach the reported performance in the paper.  

During reproduction, we made only two minimal adjustments to ensure the code could run properly (git diff results attached below):  
1. In [`step2.py`](https://github.com/YifanYang1995/GOODRL/blob/master/step2.py), we commented out the non-parallel section and uncommented the parallel code to speed up training.  
2. In [`policy/actor3.py`](https://github.com/YifanYang1995/GOODRL/blob/master/policy/actor3.py), we initialized two attributes — `entropy_count` and `grad_count` — in the `PPO` class, as missing these caused runtime attribute errors.  

<details>
<summary>Git diff result of step2.py</summary>

```diff
diff --git a/step2.py b/step2.py
index 7e557ac..8597830 100644
--- a/step2.py
+++ b/step2.py
@@ -4,7 +4,7 @@ import time, random
 import numpy as np
 from config.Params import configs
 from env.workflow_scheduling_v3.lib.poissonSampling import sample_poisson_shape
-# from joblib import Parallel, delayed
+from joblib import Parallel, delayed
 
 from policy.actor3 import PPO, BatchGraph, Memory, RolloutBuffer
 from env.workflow_scheduling_v3.simulator_wf import WFEnv
@@ -156,14 +156,14 @@ def main():
     configs.train_dataset = np.random.randint(0,wf_types,(configs.max_updates+configs.warmup_critic+1, 1, configs.wf_num ))    
     configs.arr_times_train = sample_poisson_shape(configs.arr_rate, configs.train_dataset.shape)
 
-    # non-parallel
-    meanFlowTimes = []
-    for t in range(configs.valid_num):
-        meanFlowTime = validation_H(t, configs)
-        meanFlowTimes.append(meanFlowTime)
+    # # non-parallel
+    # meanFlowTimes = []
+    # for t in range(configs.valid_num):
+    #     meanFlowTime = validation_H(t, configs)
+    #     meanFlowTimes.append(meanFlowTime)
 
-    # # parallel
-    # meanFlowTimes = Parallel(n_jobs=-1)(delayed(validation_H)( t, configs ) for t in range(configs.valid_num)) 
+    # parallel
+    meanFlowTimes = Parallel(n_jobs=configs.valid_num)(delayed(validation_H)( t, configs ) for t in range(configs.valid_num)) 
 
     t1 = time.time()
     print('Vlidation at HEFT: mean_flowtime_deterministic: {:.6f}+/-{:.6f}\t time_elapsed: {:.6f}'.\
@@ -188,18 +188,18 @@ def main():
                         {'params': algos.actor.parameters(), 'lr': configs.lr_a},
                         {'params': algos.critic.parameters(), 'lr': configs.lr_c}])
 
-    # non-parallel
-    meanFlowTimes = []
-    meanEntropies = []
-    for t in range(configs.valid_num):
-        meanFlowTime, meanEntro = validation_Org(t, configs, algos.actor, True)
-        meanFlowTimes.append(meanFlowTime)
-        meanEntropies.append(meanEntro)
+    # # non-parallel
+    # meanFlowTimes = []
+    # meanEntropies = []
+    # for t in range(configs.valid_num):
+    #     meanFlowTime, meanEntro = validation_Org(t, configs, algos.actor, True)
+    #     meanFlowTimes.append(meanFlowTime)
+    #     meanEntropies.append(meanEntro)
 
-    # # parallel
-    # valids = Parallel(n_jobs=-1)(delayed(validation_Org)( t, configs, algos.actor, True ) for t in range(configs.valid_num)) 
-    # valids = np.array(valids)
-    # meanFlowTimes,meanEntropies = valids[:,0], valids[:,1]
+    # parallel
+    valids = Parallel(n_jobs=configs.valid_num)(delayed(validation_Org)( t, configs, algos.actor, True ) for t in range(configs.valid_num)) 
+    valids = np.array(valids)
+    meanFlowTimes,meanEntropies = valids[:,0], valids[:,1]
 
     if np.mean(meanFlowTimes) < record:
         record = deepcopy(np.mean(meanFlowTimes))
@@ -216,27 +216,27 @@ def main():
 
         algos.update_idx = i_update+1 
 
-        # non-parallel
-        ep_returns = []
-        memories = []
-        for i in range(configs.num_envs):
-            memory, returni = collect_whole_rollouts(algos.actor, configs, i_update)#i)
-            temp_memory = Memory(len(memory[0]))
-            temp_memory.update(*memory)
-            if configs.gae_lambda == 1:        
-                period_returns = compute_returns(temp_memory.reward_mb)
-                buffer = RolloutBuffer(temp_memory, period_returns, period_returns, len(period_returns))
-            else:
-                temp_vals = compute_values(algos.critic, temp_memory.v_state_mb)
-                period_returns, period_adv = temp_memory.compute_gae(temp_vals)
-                buffer = RolloutBuffer(temp_memory, period_returns, period_adv, len(period_returns))
-            ep_returns.append(returni)
-            memories.append(deepcopy(buffer))
-
-        # # parallel
-        # outputs = Parallel(n_jobs=1)(delayed(collect_parallel)( i_update, algos.actor, algos.critic ) for _ in range(configs.num_envs)) 
-        # ep_returns = [item[0] for item in outputs]
-        # memories = [item[1] for item in outputs]
+        # # non-parallel
+        # ep_returns = []
+        # memories = []
+        # for i in range(configs.num_envs):
+        #     memory, returni = collect_whole_rollouts(algos.actor, configs, i_update)#i)
+        #     temp_memory = Memory(len(memory[0]))
+        #     temp_memory.update(*memory)
+        #     if configs.gae_lambda == 1:        
+        #         period_returns = compute_returns(temp_memory.reward_mb)
+        #         buffer = RolloutBuffer(temp_memory, period_returns, period_returns, len(period_returns))
+        #     else:
+        #         temp_vals = compute_values(algos.critic, temp_memory.v_state_mb)
+        #         period_returns, period_adv = temp_memory.compute_gae(temp_vals)
+        #         buffer = RolloutBuffer(temp_memory, period_returns, period_adv, len(period_returns))
+        #     ep_returns.append(returni)
+        #     memories.append(deepcopy(buffer))
+
+        # parallel
+        outputs = Parallel(n_jobs=configs.num_envs)(delayed(collect_parallel)( i_update, algos.actor, algos.critic ) for _ in range(configs.num_envs)) 
+        ep_returns = [item[0] for item in outputs]
+        memories = [item[1] for item in outputs]
 
         new_buffer = memories[0]
         new_buffer.merge(memories[1:])
@@ -255,18 +255,18 @@ def main():
 
         if (i_update + 1) % configs.log_interval == 0:
 
-            # non-parallel
-            meanFlowTimes = []
-            meanEntropies = []
-            for t in range(configs.valid_num):
-                meanFlowTime, meanEntro = validation_Org( t, configs, algos.actor, True)
-                meanFlowTimes.append(meanFlowTime)
-                meanEntropies.append(meanEntro)
-
-            # # parallel
-            # valids = Parallel(n_jobs=-1)(delayed(validation_Org)( t, configs, algos.actor, True ) for t in range(configs.valid_num)) 
-            # valids = np.array(valids)
-            # meanFlowTimes,meanEntropies = valids[:,0], valids[:,1]     
+            # # non-parallel
+            # meanFlowTimes = []
+            # meanEntropies = []
+            # for t in range(configs.valid_num):
+            #     meanFlowTime, meanEntro = validation_Org( t, configs, algos.actor, True)
+            #     meanFlowTimes.append(meanFlowTime)
+            #     meanEntropies.append(meanEntro)
+
+            # parallel
+            valids = Parallel(n_jobs=configs.valid_num)(delayed(validation_Org)( t, configs, algos.actor, True ) for t in range(configs.valid_num)) 
+            valids = np.array(valids)
+            meanFlowTimes,meanEntropies = valids[:,0], valids[:,1]     
 
             if np.mean(meanFlowTimes) < record:
                 torch.save(algos.actor.state_dict(), './logs/a_{}_{}_{}.pth'.format(configs.vm_types, configs.each_vm_type_num, configs.arr_rate) )

```

</details>

<details>
<summary>Git diff result of policy/actor3.py</summary>

```diff
diff --git a/policy/actor3.py b/policy/actor3.py
index 89cbfe2..610c0f5 100644
--- a/policy/actor3.py
+++ b/policy/actor3.py
@@ -811,6 +811,11 @@ class PPO:
 
         self.optimizer_actor = torch.optim.Adam(self.actor.parameters(), lr=configs.lr_a)
         self.optimizer_critic = torch.optim.Adam(self.critic.parameters(), lr=configs.lr_c)
+        
+        self.entropy_count = 0
+        self.grad_count = 0
 
     def train_HEFT(self, bufferdata):
         # state_mb, action_mb, reward_mb

```

</details>

Below are the key package versions from our environment, consistent with the [README](https://github.com/YifanYang1995/GOODRL/blob/master/README.md) requirements:  

<details>
<summary>Environment Details</summary>

```text
Python 3.11.13

torch==2.8.0
torch-geometric==2.6.1
rl_zoo3==2.7.0
deap==1.4.3
gym==0.26.2
joblib==1.5.2
openpyxl==3.1.5
```

</details>

Could you please check if any of the above modifications might affect the results, or if there might be an issue with the code/checkpoints currently in the repository?  

Thank you again for your kind assistance and for maintaining the project!
